File: apps/backend/internal/crypto/hybrid/hybrid.go
package hybrid

import (
	"encoding/base64"
	"fmt"
)

// HybridSigner combines Ed25519 and ML-DSA-65 for defense-in-depth.
//
// During the post-quantum transition period, hybrid signatures provide
// security against both classical and quantum attackers:
// - If quantum computers never materialize: Ed25519 remains secure
// - If quantum computers break Ed25519: ML-DSA-65 remains secure
// - Both signatures must verify for the hybrid to be valid
type HybridSigner struct {
	classical *Ed25519Signer
	pqc       *MLDSASigner
}

// NewHybridSigner generates new key pairs for both algorithms.
func NewHybridSigner() (*HybridSigner, error) {
	classical, err := NewEd25519Signer()
	if err != nil {
		return nil, fmt.Errorf("failed to create Ed25519 signer: %w", err)
	}

	pqc, err := NewMLDSASigner()
	if err != nil {
		return nil, fmt.Errorf("failed to create ML-DSA signer: %w", err)
	}

	return &HybridSigner{
		classical: classical,
		pqc:       pqc,
	}, nil
}

// NewHybridSignerFromKeys creates a hybrid signer from existing keys.
func NewHybridSignerFromKeys(keyPair *HybridKeyPair) (*HybridSigner, error) {
	classical, err := NewEd25519SignerFromKeys(keyPair.ClassicalPublic, keyPair.ClassicalPrivate)
	if err != nil {
		return nil, fmt.Errorf("failed to create Ed25519 signer: %w", err)
	}

	pqc, err := NewMLDSASignerFromKeys(keyPair.PQCPublic, keyPair.PQCPrivate)
	if err != nil {
		return nil, fmt.Errorf("failed to create ML-DSA signer: %w", err)
	}

	return &HybridSigner{
		classical: classical,
		pqc:       pqc,
	}, nil
}

// Algorithm returns the algorithm identifier.
func (s *HybridSigner) Algorithm() Algorithm {
	return AlgorithmHybrid
}

// PublicKey returns the combined public keys as a base64-encoded hybrid format.
// Format: base64(classical_pub || pqc_pub)
func (s *HybridSigner) PublicKey() string {
	classical := s.classical.PublicKeyBytes()
	pqc := s.pqc.PublicKeyBytes()

	combined := make([]byte, len(classical)+len(pqc))
	copy(combined[:len(classical)], classical)
	copy(combined[len(classical):], pqc)

	return base64.StdEncoding.EncodeToString(combined)
}

// PublicKeyBytes returns the combined public key bytes.
func (s *HybridSigner) PublicKeyBytes() []byte {
	classical := s.classical.PublicKeyBytes()
	pqc := s.pqc.PublicKeyBytes()

	combined := make([]byte, len(classical)+len(pqc))
	copy(combined[:len(classical)], classical)
	copy(combined[len(classical):], pqc)

	return combined
}

// ClassicalPublicKey returns the Ed25519 public key.
func (s *HybridSigner) ClassicalPublicKey() string {
	return s.classical.PublicKey()
}

// PQCPublicKey returns the ML-DSA-65 public key.
func (s *HybridSigner) PQCPublicKey() string {
	return s.pqc.PublicKey()
}

// KeyPair returns the hybrid key pair for storage.
func (s *HybridSigner) KeyPair() *HybridKeyPair {
	return &HybridKeyPair{
		Algorithm:        AlgorithmHybrid,
		ClassicalPublic:  s.classical.PublicKey(),
		ClassicalPrivate: s.classical.PrivateKey(),
		PQCPublic:        s.pqc.PublicKey(),
		PQCPrivate:       s.pqc.PrivateKey(),
	}
}

// IsSimulated returns true if the PQC component is using placeholder implementation.
func (s *HybridSigner) IsSimulated() bool {
	return s.pqc.IsSimulated()
}

// Sign signs a message with both algorithms and returns the hybrid signature.
func (s *HybridSigner) Sign(message []byte) (string, error) {
	sig, err := s.SignBytes(message)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(sig), nil
}

// SignBytes signs a message with both algorithms and returns encoded hybrid signature.
func (s *HybridSigner) SignBytes(message []byte) ([]byte, error) {
	// Sign with classical algorithm
	classicalSig, err := s.classical.SignBytes(message)
	if err != nil {
		return nil, fmt.Errorf("classical signing failed: %w", err)
	}

	// Sign with PQC algorithm
	pqcSig, err := s.pqc.SignBytes(message)
	if err != nil {
		return nil, fmt.Errorf("PQC signing failed: %w", err)
	}

	// Encode as hybrid signature
	hybrid := &HybridSignature{
		Classical: classicalSig,
		PQC:       pqcSig,
	}

	return EncodeHybridSignature(hybrid), nil
}

// Verify verifies both signatures in a hybrid signature.
// BOTH must be valid for the hybrid signature to be valid.
func (s *HybridSigner) Verify(message []byte, signatureB64 string) (bool, error) {
	signature, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return false, fmt.Errorf("failed to decode signature: %w", err)
	}
	return s.VerifyBytes(message, signature)
}

// VerifyBytes verifies both signatures in a hybrid signature.
func (s *HybridSigner) VerifyBytes(message []byte, signature []byte) (bool, error) {
	// Decode hybrid signature
	hybrid, err := DecodeHybridSignature(signature)
	if err != nil {
		return false, fmt.Errorf("failed to decode hybrid signature: %w", err)
	}

	// Verify classical signature
	classicalValid, err := s.classical.VerifyBytes(message, hybrid.Classical)
	if err != nil {
		return false, fmt.Errorf("classical verification failed: %w", err)
	}
	if !classicalValid {
		return false, nil
	}

	// Verify PQC signature
	pqcValid, err := s.pqc.VerifyBytes(message, hybrid.PQC)
	if err != nil {
		return false, fmt.Errorf("PQC verification failed: %w", err)
	}

	// Both must be valid
	return classicalValid && pqcValid, nil
}

// HybridVerifier verifies hybrid signatures without private key access.
type HybridVerifier struct {
	classical *Ed25519Verifier
	pqc       *MLDSAVerifier
}

// NewHybridVerifier creates a verifier from both public keys.
func NewHybridVerifier(classicalPublicB64, pqcPublicB64 string) (*HybridVerifier, error) {
	classical, err := NewEd25519VerifierFromPublicKey(classicalPublicB64)
	if err != nil {
		return nil, fmt.Errorf("failed to create Ed25519 verifier: %w", err)
	}

	pqc, err := NewMLDSAVerifierFromPublicKey(pqcPublicB64)
	if err != nil {
		return nil, fmt.Errorf("failed to create ML-DSA verifier: %w", err)
	}

	return &HybridVerifier{
		classical: classical,
		pqc:       pqc,
	}, nil
}

// Algorithm returns the algorithm identifier.
func (v *HybridVerifier) Algorithm() Algorithm {
	return AlgorithmHybrid
}

// Verify verifies both signatures in a hybrid signature.
func (v *HybridVerifier) Verify(message []byte, signatureB64 string) (bool, error) {
	signature, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return false, fmt.Errorf("failed to decode signature: %w", err)
	}
	return v.VerifyBytes(message, signature)
}

// VerifyBytes verifies both signatures in a hybrid signature.
func (v *HybridVerifier) VerifyBytes(message []byte, signature []byte) (bool, error) {
	// Decode hybrid signature
	hybrid, err := DecodeHybridSignature(signature)
	if err != nil {
		return false, fmt.Errorf("failed to decode hybrid signature: %w", err)
	}

	// Verify classical signature
	classicalValid, err := v.classical.VerifyBytes(message, hybrid.Classical)
	if err != nil {
		return false, fmt.Errorf("classical verification failed: %w", err)
	}
	if !classicalValid {
		return false, nil
	}

	// Verify PQC signature
	pqcValid, err := v.pqc.VerifyBytes(message, hybrid.PQC)
	if err != nil {
		return false, fmt.Errorf("PQC verification failed: %w", err)
	}

	// Both must be valid
	return classicalValid && pqcValid, nil
}

// SignatureInfo contains details about signature verification.
type SignatureInfo struct {
	Algorithm          Algorithm `json:"algorithm"`
	ClassicalValid     bool      `json:"classical_valid"`
	PQCValid           bool      `json:"pqc_valid"`
	ClassicalAlgorithm string    `json:"classical_algorithm"`
	PQCAlgorithm       string    `json:"pqc_algorithm"`
	SignatureSize      int       `json:"signature_size"`
}

// VerifyWithInfo verifies and returns detailed information.
func (v *HybridVerifier) VerifyWithInfo(message []byte, signatureB64 string) (*SignatureInfo, error) {
	signature, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return nil, fmt.Errorf("failed to decode signature: %w", err)
	}

	// Decode hybrid signature
	hybrid, err := DecodeHybridSignature(signature)
	if err != nil {
		return nil, fmt.Errorf("failed to decode hybrid signature: %w", err)
	}

	info := &SignatureInfo{
		Algorithm:          AlgorithmHybrid,
		ClassicalAlgorithm: string(AlgorithmEd25519),
		PQCAlgorithm:       string(AlgorithmMLDSA65),
		SignatureSize:      len(signature),
	}

	// Verify classical
	classicalValid, _ := v.classical.VerifyBytes(message, hybrid.Classical)
	info.ClassicalValid = classicalValid

	// Verify PQC
	pqcValid, _ := v.pqc.VerifyBytes(message, hybrid.PQC)
	info.PQCValid = pqcValid

	return info, nil
}

File: sdk/python/aim_sdk/crypto/hybrid.py
"""
Hybrid signing combining Ed25519 and ML-DSA-65.

This provides defense-in-depth during the post-quantum transition period:
- If quantum computers never materialize: Ed25519 remains secure
- If quantum computers break Ed25519: ML-DSA-65 remains secure
- Both signatures must verify for the hybrid to be valid
"""

import base64
from dataclasses import dataclass
from typing import Optional

from .signer import (
    Algorithm,
    Signer,
    Verifier,
    HybridKeyPair,
    HybridSignature,
    encode_hybrid_signature,
    decode_hybrid_signature,
)
from .ed25519 import Ed25519Signer, Ed25519Verifier
from .mldsa import MLDSASigner, MLDSAVerifier


@dataclass
class SignatureInfo:
    """Detailed information about signature verification."""

    algorithm: Algorithm
    classical_valid: bool
    pqc_valid: bool
    classical_algorithm: str
    pqc_algorithm: str
    signature_size: int


class HybridSigner(Signer):
    """
    Hybrid signer combining Ed25519 and ML-DSA-65.

    Provides defense-in-depth: both signatures must verify.
    """

    def __init__(self, classical: Ed25519Signer, pqc: MLDSASigner):
        """
        Initialize hybrid signer with both component signers.

        Args:
            classical: Ed25519 signer
            pqc: ML-DSA-65 signer
        """
        self._classical = classical
        self._pqc = pqc

    @classmethod
    def generate(cls) -> "HybridSigner":
        """
        Generate new key pairs for both algorithms.

        Returns:
            New HybridSigner with generated keys
        """
        classical = Ed25519Signer.generate()
        pqc = MLDSASigner.generate()
        return cls(classical, pqc)

    @classmethod
    def from_key_pair(cls, key_pair: HybridKeyPair) -> "HybridSigner":
        """
        Create hybrid signer from an existing key pair.

        Args:
            key_pair: HybridKeyPair with keys for both algorithms

        Returns:
            HybridSigner instance
        """
        classical = Ed25519Signer.from_keys(
            key_pair.classical_public,
            key_pair.classical_private,
        )
        pqc = MLDSASigner.from_keys(
            key_pair.pqc_public,
            key_pair.pqc_private,
        )
        return cls(classical, pqc)

    @property
    def algorithm(self) -> Algorithm:
        return Algorithm.HYBRID

    @property
    def is_simulated(self) -> bool:
        """Return True if PQC component is using placeholder implementation."""
        return self._pqc.is_simulated

    def public_key(self) -> str:
        """
        Return combined public keys as base64-encoded hybrid format.

        Format: base64(classical_pub || pqc_pub)
        """
        combined = self.public_key_bytes()
        return base64.b64encode(combined).decode("utf-8")

    def public_key_bytes(self) -> bytes:
        """Return combined public key bytes."""
        classical = self._classical.public_key_bytes()
        pqc = self._pqc.public_key_bytes()
        return classical + pqc

    def classical_public_key(self) -> str:
        """Return the Ed25519 public key."""
        return self._classical.public_key()

    def pqc_public_key(self) -> str:
        """Return the ML-DSA-65 public key."""
        return self._pqc.public_key()

    def key_pair(self) -> HybridKeyPair:
        """Return the hybrid key pair for storage."""
        return HybridKeyPair(
            algorithm=Algorithm.HYBRID,
            classical_public=self._classical.public_key(),
            classical_private=self._classical.private_key(),
            pqc_public=self._pqc.public_key(),
            pqc_private=self._pqc.private_key(),
        )

    def sign(self, message: bytes) -> str:
        """
        Sign message with both algorithms and return base64-encoded hybrid signature.

        Args:
            message: The message to sign

        Returns:
            Base64-encoded hybrid signature
        """
        sig = self.sign_bytes(message)
        return base64.b64encode(sig).decode("utf-8")

    def sign_bytes(self, message: bytes) -> bytes:
        """
        Sign message with both algorithms and return encoded hybrid signature.

        Args:
            message: The message to sign

        Returns:
            Encoded hybrid signature bytes
        """
        # Sign with classical algorithm
        classical_sig = self._classical.sign_bytes(message)

        # Sign with PQC algorithm
        pqc_sig = self._pqc.sign_bytes(message)

        # Encode as hybrid signature
        hybrid = HybridSignature(classical=classical_sig, pqc=pqc_sig)
        return encode_hybrid_signature(hybrid)

    def verify(self, message: bytes, signature: str) -> bool:
        """
        Verify both signatures in a hybrid signature.

        BOTH must be valid for the hybrid signature to be valid.

        Args:
            message: The original message
            signature: Base64-encoded hybrid signature

        Returns:
            True if BOTH signatures are valid
        """
        try:
            sig_bytes = base64.b64decode(signature)
            return self.verify_bytes(message, sig_bytes)
        except Exception:
            return False

    def verify_bytes(self, message: bytes, signature: bytes) -> bool:
        """
        Verify both signatures in a hybrid signature.

        Args:
            message: The original message
            signature: Encoded hybrid signature bytes

        Returns:
            True if BOTH signatures are valid
        """
        try:
            # Decode hybrid signature
            hybrid = decode_hybrid_signature(signature)

            # Verify classical signature
            if not self._classical.verify_bytes(message, hybrid.classical):
                return False

            # Verify PQC signature
            if not self._pqc.verify_bytes(message, hybrid.pqc):
                return False

            return True
        except Exception:
            return False


class HybridVerifier(Verifier):
    """
    Hybrid verifier without private key access.

    Verifies both Ed25519 and ML-DSA-65 signatures.
    """

    def __init__(self, classical: Ed25519Verifier, pqc: MLDSAVerifier):
        """
        Initialize hybrid verifier with both component verifiers.

        Args:
            classical: Ed25519 verifier
            pqc: ML-DSA-65 verifier
        """
        self._classical = classical
        self._pqc = pqc

    @classmethod
    def from_public_keys(
        cls, classical_public_b64: str, pqc_public_b64: str
    ) -> "HybridVerifier":
        """
        Create hybrid verifier from both public keys.

        Args:
            classical_public_b64: Base64-encoded Ed25519 public key
            pqc_public_b64: Base64-encoded ML-DSA-65 public key

        Returns:
            HybridVerifier instance
        """
        classical = Ed25519Verifier.from_public_key(classical_public_b64)
        pqc = MLDSAVerifier.from_public_key(pqc_public_b64)
        return cls(classical, pqc)

    @property
    def algorithm(self) -> Algorithm:
        return Algorithm.HYBRID

    def verify(self, message: bytes, signature: str) -> bool:
        """
        Verify both signatures in a hybrid signature.

        Args:
            message: The original message
            signature: Base64-encoded hybrid signature

        Returns:
            True if BOTH signatures are valid
        """
        try:
            sig_bytes = base64.b64decode(signature)
            return self.verify_bytes(message, sig_bytes)
        except Exception:
            return False

    def verify_bytes(self, message: bytes, signature: bytes) -> bool:
        """
        Verify both signatures in a hybrid signature.

        Args:
            message: The original message
            signature: Encoded hybrid signature bytes

        Returns:
            True if BOTH signatures are valid
        """
        try:
            # Decode hybrid signature
            hybrid = decode_hybrid_signature(signature)

            # Verify classical signature
            if not self._classical.verify_bytes(message, hybrid.classical):
                return False

            # Verify PQC signature
            if not self._pqc.verify_bytes(message, hybrid.pqc):
                return False

            return True
        except Exception:
            return False

    def verify_with_info(self, message: bytes, signature: str) -> SignatureInfo:
        """
        Verify and return detailed information about the signature.

        Args:
            message: The original message
            signature: Base64-encoded hybrid signature

        Returns:
            SignatureInfo with details about each component
        """
        sig_bytes = base64.b64decode(signature)
        hybrid = decode_hybrid_signature(sig_bytes)

        classical_valid = self._classical.verify_bytes(message, hybrid.classical)
        pqc_valid = self._pqc.verify_bytes(message, hybrid.pqc)

        return SignatureInfo(
            algorithm=Algorithm.HYBRID,
            classical_valid=classical_valid,
            pqc_valid=pqc_valid,
            classical_algorithm=str(Algorithm.ED25519.value),
            pqc_algorithm=str(Algorithm.MLDSA65.value),
            signature_size=len(sig_bytes),
        )
